.github/workflows/black.yml

```diff
@@ -1,73 +0,0 @@
-name: black
```

Comment:
Should we remove this action?

Reply:
Bring this over from #1276:
> I don't think the solution to this is to disable flake8 with pre-commit.ci -- instead, since pre-commit.ci is supposedly faster than GitHub Actions/workflows, I'd rather disable the style checking workflows under our repos .github directory remove/disable them. This is especially true because we seem to be having issues with GitHub Actions and black atm @naveen521kk .

Reply:
> I'd rather disable the style checking workflows under our repos .github directory remove/disable them.

The file I talking about isn't that?

I think we should note that black doesn't have a stable release, yet! so there is no guarantee for compatibility between releases. Have a look at the time this action was created for context, why this is used. It was when a new black pre-release was released and every other check started to fail; we formated the whole code base with this action at that point in time.

Reply:
So this workflow is just to update black and then format the whole codebase with the updated black?

Why not just use the version of black specified `.pre-commit-config.yaml` and update the version when we want/need to instead of automatically updating black?

Reply:
> So this workflow is just to update black and then format the whole codebase with the updated black?

yes

> Why not just use the version of black specified .pre-commit-config.yaml and update the version when we want/need to instead of automatically updating black?

See previous discussion why it shouldn't.

> pre-commit.ci will push commits for the user using the version of black specified here; if it fails, it'll just automatically apply the formatting to the whole codebase.

No idea, how those works though, so if it works it is fine.

Reply:
Plus! pre-commit is faster:
![image](https://user-images.githubusercontent.com/29168469/113857448-635e3780-9757-11eb-96ea-973c91e2bff0.png)
vs.
#1269
![image](https://user-images.githubusercontent.com/29168469/113857591-8c7ec800-9757-11eb-8d47-9094c1644e7e.png)
![image](https://user-images.githubusercontent.com/29168469/113857630-986a8a00-9757-11eb-9bd8-59f75b44dd54.png)

Reply:
>No idea, how those works though, so if it works it is fine.

Me neither. But @kolibril13 did some testing in a repo and it seems to intercept the commit, apply stuff and then re-commit. See his PRs in his test repo: https://github.com/kolibril13/okapi/pulls

Reply:
I thought it was just going to reproduce what these actions did. I don't like someone else even if it a bot to commit in my branch, cause later I may not pull and it will create a lot of merge conflict, which would waste time actually. I should be able to disable this bot for my PR only then this needs to be added to this repo.

Reply:
I'm not sure if there's a way to disable pre-commit.ci from committing on a user/PR basis. I'm also not sure how bad handling merge-conflicts would be if someone forgot to pull -- but the idea should be that if the user develops with the `pre-commit` githook installed, they won't have to worry about merge-conflicts anyways!

Reply:
I agree with you in principle; I also feel it a bit invasive for a bot to directly commit to my branch. However, since we have introduced the requirement that branches need to include the latest version of `master` before they can be merged (and the, I have to admit, pretty convenient `Merge master` button in the interface that everyone with write permissions can use), we sort of already have the situation that your local branch and the remove can diverge.

If you want to disable that, you can always uncheck the "Allow edits from maintainers" box -- at the cost of noone else being able to push to your branch except yourself. Also, the bot won't commit if you have already formatted your changes accordingly (which would be ensured if you would use the pre-commit hook).

Overall, I prefer more convenience for contributors and reviewers (it happened more than once that I didn't write black-compliant code when adding suggestions, and that would usually require another iteration before such a PR is then ready to be merged...) -- especially considering the large amount of PRs that this repo has received recently. (>40!!)

Reply:
After I've asked him to do so, @behackl  installed the pre-commit.ci.
Here is how I see the situation: We decided that we want to enforce isort, black and flake8 in our repo, and sometimes it might happen that people forget to run black/isort, or they change code from the code suggestion. Then this had to be done manually. But as there was the idea of automating this process, we came up to think about a tool that can make life easier for us, which is in this case pre-commit.ci.
> cause later I may not pull and it will create a lot of merge conflict

As already mentioned by @jsonvillanueva , this case is rather unlikely, as all pushes will already have black+isort.
And when there was a code suggestion, making another pull is required eighter way.
Also, I don't think that it would be too much effort to merge these conflicts from white spacing.

And lastly, in the end, every pr *has* to have these formatting to be merged into master. So for me, the benefits clearly outweigh the drawbacks.

Reply:
The `Merge Master` button works very differently than this bot itself. That button just merges a branch which I am sure no one will do unless it is final merging time. Also, I can forget to pull it and there won't be any merge conflict(merge conflict happens very rarely), but this isn't the case with the bot as it touches the Code which I worked on in previous commits.

With that said I am thinking about the situation of people who are new to git itself(I am thinking about myself when I was getting started with manim), this bot can cause a lot of problems with them, because they don't know about fixing merge conflicts and could have a hard time working with that and it could make them not contribute in a way. I would say there would "less convenience for contributors" because of this.

Reply:
But remember, when we introduced pre-commit we made it part of the contributing guidelines [for first-time contributors section](https://docs.manim.community/en/latest/contributing/development.html#for-first-time-contributors):
![image](https://user-images.githubusercontent.com/29168469/113863058-0ade6880-975e-11eb-9184-e99c384830e8.png)
If they don't install the git-hook, they've really brought the merge-conflicts on themselves...

Reply:
I would rather like a bot that will do these only when I or anybody with write access say it to. Maybe find something like that, rather than this?

> as all pushes will already have black+isort.

I would only like to add these cosmetics to the code only when the PR is about to be merged. Or else, I won't be doing this.

>  Also, I don't think that it would be too much effort to merge these conflicts from white spacing.

I think you haven't experienced it and it could be confusing and dreading.

Reply:
> this bot can cause a lot of problems with them, because they don't know about fixing merge conflicts and could have a hard time working with that and it could make them not contribute in a way.

I also remember that it was for me tricky to get started with git. But as @behackl mentioned, saving one iteration for all prs is a HUGE benefit.
And I just thought about the situation with the merge conflict:
I think in that case we can tell the users in the docs contribution guide that they can just take all the local changes and discard the remote changes, as the bot will run in the end again.
To consider your concerns I have some ideas:
* We can put some effort to add this bot behaviour to the docs contribution guide.
* Maybe we can even make an issue at pre-commit.ci and ask them if it is possible to make a second Button on the git pr page "Apply Code Formatting", which would trigger this action only if the user presses it actively.

Reply:
> which I am sure no one will do unless it is final merging time

That's not quite the case if you look over the PRs from the last few days. Just because someone used the button it doesn't mean that the PR got merged soon after that; sometimes additional development happened.

> and there won't be any merge conflict (merge conflict happens very rarely)

They happen rarely, but they still happen.

> this bot can cause a lot of problems with them, because they don't know about fixing merge conflicts and could have a hard time working with that and it could make them not contribute in a way

I don't disagree.

> I would say there would "less convenience for contributors" because of this.

I disagree. In the vast majority of cases it will simply help to fix silly pipeline build errors, and given that many people work with IDEs which fetch remote changes automatically once in a while, I feel that the cases where this would introduce serious problems would still be rather rare.

That being said, overall I'd also prefer a manual action that commits the fixes over an automatic one -- as long as we don't have to write our own actions for that.

Reply:
Okay, so my idea to proceed:
* I can create an issue at pre-commit.ci and ask them if they are planning to add a manual action
* We merge this pr and write a good documentation of how to use this feature (maybe even with a short video/gif?)
    * when in future some situations occur which @naveen521kk pointed out, we can come back to this conversation.

Reply:
Fine. I think until we find something which doesn't do this automatically, I think we tried another bot called "Restyled" remember, I would not be allowing it to commit to my branch.

Reply:
About my earlier comment: https://github.com/ManimCommunity/manim/pull/1278#discussion_r608588275
Should I remove the mentions of optional/recommended?

Reply:
@naveen521kk , @jsonvillanueva : The author of pre-commit.ci got back to me about the manual trigger:
https://github.com/pre-commit-ci/issues/issues/52#issuecomment-815874609